EFCore.CheckConstraints/Internal/EnumCheckConstraintConvention.cs

```diff
@@ -41,34 +58,83 @@ public virtual void ProcessModelFinalizing(IConventionModelBuilder modelBuilder,
+                    || typeMapping is null
+                    || propertyType.IsDefined(typeof(FlagsAttribute), true)
+                    || property.GetColumnName(tableIdentifier) is not { } columnName)
                 {
-                    var enumValues = Enum.GetValues(propertyType);
-                    if (enumValues.Length <= 0)
-                    {
-                        continue;
-                    }
+                    continue;
+                }
+
+                var values = Enum.GetValues(propertyType);
+                if (values.Length == 0)
+                {
+                    continue;
+                }
 
-                    sql.Clear();
+                sql.Clear();
 
-                    sql.Append(_sqlGenerationHelper.DelimitIdentifier(columnName));
+                sql.Append(_sqlGenerationHelper.DelimitIdentifier(columnName));
+
+                if (TryParseContiguousRange(typeMapping, values, out var minValue, out var maxValue))
+                {
+                    sql.Append(" BETWEEN ");
+                    sql.Append(minValue);
+                    sql.Append(" AND ");
+                    sql.Append(maxValue);
+                }
+                else
+                {
                     sql.Append(" IN (");
-                    foreach (var item in enumValues)
+                    foreach (var item in values)
                     {
                         var value = typeMapping.GenerateSqlLiteral(item);
                         sql.Append($"{value}, ");
                     }
 
                     sql.Remove(sql.Length - 2, 2);
-                    sql.Append(")");
-
-                    var constraintName = $"CK_{tableName}_{columnName}_Enum";
-                    entityType.AddCheckConstraint(constraintName, sql.ToString());
+                    sql.Append(')');
                 }
+
+                var constraintName = $"CK_{tableName}_{columnName}_Enum";
+                entityType.AddCheckConstraint(constraintName, sql.ToString());
             }
         }
     }
+
+    private bool TryParseContiguousRange(CoreTypeMapping typeMapping, IEnumerable values, out object? minValue, out object? maxValue)
+    {
+        minValue = default;
+        maxValue = default;
+
+        if (typeMapping.Converter?.ProviderClrType is null || !_knownTypes.Contains(typeMapping.Converter.ProviderClrType))
+        {
+            return false;
+        }
+
+        var underlyingType = Enum.GetUnderlyingType(typeMapping.ClrType);
+
+        var parameters = new[] { values, minValue, maxValue };
+
+        var success = GetType().GetTypeInfo().GetDeclaredMethod(nameof(TryGetMinMax))!.MakeGenericMethod(underlyingType)
+            .Invoke(null, parameters)!;
+
+        minValue = parameters[1];
+        maxValue = parameters[2];
+
+        return (bool)success;
+    }
+
+    private static bool TryGetMinMax<T>(IEnumerable values, out T minValue, out T maxValue)
+        where T : IBinaryInteger<T>, new()
```

Comment:
Any specific reason to use IBinaryInteger rather than just INumber? You don't seem to be using any functionality specific to it...

Reply:
Not really. I just guessed based on the documentation but `INumber` works too.